src/geometry/manifold/algebra/left_invariant_derivation.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import geometry.manifold.derivation_bundle
+
+/-!
+
+# Left invariant derivations
+
+In this file we define the concept of left invariant derivation for a Lie group. The concept is
+analogous to the more classical concept of left invariant vector fields, and it holds that the
+derivation associated to a vector field is left invariant iff the the field is.
+
+Moreover we prove that `left_invariant_derivation I G` has the structure of a Lie algebra, hence
+implementing one of the possible definitions of Lie algebra.
+
+-/
+
+noncomputable theory
+
+open_locale lie_group manifold derivation
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+(G : Type*) [topological_space G] [charted_space H G] [monoid G] [has_smooth_mul I G] (g h : G)
+
+-- Generate trivial has_sizeof instance. It prevents weird type class inference timeout problems
+local attribute [nolint instance_priority, instance, priority 10000]
+private def disable_has_sizeof {α} : has_sizeof α := ⟨λ _, 0⟩
+
+/--
+Left-invariant global derivations.
+
+A global derivation is left-invariant if it is equal to its pullback along left multiplication by
+an arbitrary element of `G`.
+-/
+structure left_invariant_derivation extends derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯ :=
+(left_invariant'' : ∀ f : C^∞⟮I, G; 𝕜⟯, ∀ g, 𝒅(𝑳 I g) 1 (derivation.eval_at 1 to_derivation) f =
+  derivation.eval_at g to_derivation f)
+
+variables {I G}
+
+namespace left_invariant_derivation
+
+instance : has_coe (left_invariant_derivation I G) (derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯)
+:= ⟨λ X, X.to_derivation⟩
+
+instance : has_coe_to_fun (left_invariant_derivation I G) := ⟨_, λ X, X.to_derivation.to_fun⟩
+
+variables
+{M : Type*} [topological_space M] [charted_space H M] {x : M} {r : 𝕜}
+{X Y : left_invariant_derivation I G} {f f' : C^∞⟮I, G; 𝕜⟯}
+
+lemma to_fun_eq_coe : X.to_fun = ⇑X := rfl
+lemma coe_fn_coe : ⇑(X : C^∞⟮I, G; 𝕜⟯ →ₗ[𝕜] C^∞⟮I, G; 𝕜⟯) = X := rfl
```

Comment:
```suggestion
@[simp] lemma to_fun_eq_coe : X.to_fun = ⇑X := rfl
@[simp] lemma coe_to_linear_map : ⇑(X : C^∞⟮I, G; 𝕜⟯ →ₗ[𝕜] C^∞⟮I, G; 𝕜⟯) = X := rfl
```

Reply:
Actually, I don't think this `coe_to_linear_map` lemma should exist at all, as you get there via `coe_derivation` and `derivation.coe_linear_map`.

Reply:
These lemmas cannot be simp. For example, the simplification don by `to_fun_eq_coe` is already carried out by `left_invariant_derivation.to_derivation_eq_coe, derivation.to_fun_eq_coe`.

Reply:
I'm surprised that ` derivation.to_fun_eq_coe` matches `to_fun_eq_coe` - I guess this is new-style structure behavior that I wasn't expecting. I'll have a go myself once a cache is available.

Reply:
Huh, turns out that when you write `X.to_fun` lean sees that as notation for `X.to_derivation.to_fun`.

Reply:
So the current one is the best way to express it right?